text/000-buildpacks-contribute-default-process-type.md

```diff
@@ -0,0 +1,168 @@
+- Changing the way things work now could surprise some people.
+- One more thing for buildpack authors to know and care about.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+  - Keep things as they are.
+  - Earlier iterations of this PR, including:
+
+    * Having `default-process-type` be a top-level key in `<layers>/launch.toml`, as opposed to a field on each process.
+    * Having a concept of `override` tied to either
+
+        (a) each process type, such that process types specified by later buildpacks MAY override process types with the same name that were specified by earlier buildpacks, depending on the value of `override`
+
+        (b) the default process type itself, such that a buildpack can indicate something like "make this the default process, but only if other buildpacks didn't specify a default process"
+
+- Why is this proposal the best? Easier for end users! It simplifies the build process (need to provide one fewer flag, don't need to rebuild if process types are not known ahead of time).
+- What is the impact of not doing this? A moderate amount of frustration.
+
+# Prior Art
+[prior-art]: #prior-art
+
+Discuss prior art, both the good and bad.
+
+`pack build` has a `--default-process` flag. If this flag is not provided by the user, `pack` will provide `-process-type web` to the lifecycle. If `web` is not a valid process type, the lifecycle will warn and set the image entrypoint to `/cnb/lifecycle/launcher`, which will require the user to specify the desired process in the command when running the app image. This has the advantage of "just working" for web apps. However for non-web apps, the user must know what process types exist and pass the desired default explicitly when building the app, otherwise they will get an image that must be configured at runtime.
+
+The lifecycle if not provided any `-process-type` will set the default process type to the only process type (if there is only one), otherwise it will warn and set the image entrypoint to `/cnb/lifecycle/launcher` as described above. This has the advantage of "just working" for single-process apps. However this requires the lifecycle to be "opinionated" and may be surprising for some users. See [spec PR](https://github.com/buildpacks/spec/pull/137) for the proposed removal of this behavior which should go hand-in-hand with this RFC.
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Should `pack` continue to pass `-process-type web`?
+
+  Options discussed:
+
+  A. `pack` stops sending `-process-type web` when `CNB_PLATFORM_API` is greater than `0.X`. The lifecycle will assume `default = true` for `web` processes from buildpacks implementing buildpack APIs earlier than `0.Y`. The lifecycle could print a warning when buildpacks do not specify a default process type. `-process-type <unknown process type>` could then become a failure case (vs. warn, what it is today).
```

Comment:
It sounds like this is the preferred option. Unless there are any objections, I can update the RFC to state that this will be the migration path.

Reply:
I agree that we should go forward with this one.

One small nitpick: Not sure if I think the lifecycle should print a _warning_ if buildpacks do not specify a default. I think we imagined "no default" may be an intentional case for some users of the new APIs and we try not to warn on things that could be intentional. We can certainly print an info statement though.

Reply:
Updated the language from warn -> info message.